protect: Allow admins to mark page protections that were requested at RfPP

molly opened the pull request:
Add a checkbox below the "reason" textarea to allow admins to mark page protections that were requested at RfPP. When the checkbox is checked, an input appears to allow them to enter the revision ID of the RfPP page where the request was made, which will then add a permalink to the protection log entry in the format `[[Special:Permalink/{REVISION_ID}#{PAGE_NAME}]]`.

An example log entry made using this feature can be seen at https://en.wikipedia.org/w/index.php?title=Special:Log&page=Mercian+Regiment.

Give a shout if anything needs changing or cleanup; this is my first time working in this project 🙂 

Closes #1247.

Amorymeltzer commented on modules/twinkleprotect.js:
```diff
@@ -568,6 +596,7 @@ Twinkle.protect.callback.changeAction = function twinkleprotectCallbackChangeAct
 
 		// reduce vertical height of dialog
 		$(e.target.form).find('fieldset[name="field2"] select').parent().css({ display: 'inline-block', marginRight: '0.5em' });
+		$(e.target.form).find('fieldset[name="field2"] input[name="protectReason_notes_rfppRevid"]').parent().css({display: 'none', marginLeft: '15px'});
```
```suggestion
		$(e.target.form).find('fieldset[name="field2"] input[name="protectReason_notes_rfppRevid"]').parent().css({display: 'inline-block', marginLeft: '15px'}).hide();
```

Mainly so we can just `.show()` and `.hide()` later.

Amorymeltzer commented on modules/twinkleprotect.js:
```diff
@@ -1349,6 +1382,38 @@ Twinkle.protect.callback.evaluate = function twinkleprotectCallbackEvaluate(e) {
+		} else {
+			Twinkle.protect.protectReasonAnnotations = [];
+			form.protectReason_notes_rfppRevid.value = '';
+			$(form.protectReason_notes_rfppRevid).parent().css({display: 'none'});
```
```suggestion
			$(form.protectReason_notes_rfppRevid).parent().hide();
```

Amorymeltzer commented on modules/twinkleprotect.js:
```diff
@@ -1349,6 +1382,38 @@ Twinkle.protect.callback.evaluate = function twinkleprotectCallbackEvaluate(e) {
+Twinkle.protect.protectReasonAnnotations = [];
+Twinkle.protect.callback.annotateProtectReason = function twinkleprotectCallbackAnnotateProtectReason(e) {
+	var form = e.target.form;
+	var protectReason = form.protectReason.value.replace('; ' + Twinkle.protect.protectReasonAnnotations.join(': '), '');
```
As noted above:

```suggestion
	var protectReason = form.protectReason.value.replace(new RegExp('(?:; )?' + mw.util.escapeRegExp(Twinkle.protect.protectReasonAnnotations.join(': '))), '');
```

Amorymeltzer commented on modules/twinkleprotect.js:
```diff
@@ -1349,6 +1382,38 @@ Twinkle.protect.callback.evaluate = function twinkleprotectCallbackEvaluate(e) {
+	if (!Twinkle.protect.protectReasonAnnotations.length) {
+		form.protectReason.value = protectReason;
+	} else {
+		form.protectReason.value = protectReason + '; ' + Twinkle.protect.protectReasonAnnotations.join(': ');
```
As noted above:
```suggestion
		form.protectReason.value = (protectReason ? protectReason + '; ' : '') + Twinkle.protect.protectReasonAnnotations.join(': ');
```

Amorymeltzer commented on modules/twinkleprotect.js:
```diff
@@ -1349,6 +1382,38 @@ Twinkle.protect.callback.evaluate = function twinkleprotectCallbackEvaluate(e) {
+			return el.indexOf('[[Special:Permalink') === -1;
+		});
+		if (e.target.value.length) {
+			var pageName = Morebits.pageNameNorm;
+			var permalink = '[[Special:Permalink/' + e.target.value + '#' + pageName + ']]';
```
```suggestion
			var permalink = '[[Special:Permalink/' + e.target.value + '#' + Morebits.pageNameNorm + ']]';
```

Amorymeltzer reviewed (commented):
Nice @molly, this looks pretty great to me!  I've got a couple of mostly minor suggestions, the only real notable one is handling the replacement.  Pretty sleek.

molly commented on modules/twinkleprotect.js:
```diff
@@ -471,6 +471,34 @@ Twinkle.protect.callback.changeAction = function twinkleprotectCallbackChangeAct
+				label: 'RfPP revision ID',
+				name: 'protectReason_notes_rfppRevid',
+				value: '',
+				tooltip: 'Optional revision ID of the RfPP page where protection was requested'
```
Yeah, that was my thought as well (that sysops ought to know how to find a revID).

molly commented:
I've addressed all your PR feedback, which I've left in a separate commit in case you want to give it another once-over. Let me know if you want me to squash the commits!

Amorymeltzer commented on modules/twinkleprotect.js:
```diff
@@ -1026,8 +1055,12 @@ Twinkle.protect.callback.changePreset = function twinkleprotectCallbackChangePre
+		var annotations = '';
+		if (Twinkle.protect.protectReasonAnnotations.length) {
+			annotations = '; ' + Twinkle.protect.protectReasonAnnotations.join(': ');
+		}
```
Hah, I had a bug in here thanks to gh-flavored md's nicety, hope it didn't cause any confusion!

Amorymeltzer reviewed (approved):
Nice, LGTM @molly!  Hope I didn't seem like I was insisting on the language, just expressing a preference.  You can squash if you like, or I can take care of it, nbd.

Curious what you think about adding a validation that the revid at least looks right?  That is, something in `Twinkle.protect.callback.evaluate` that does `input.protectReason_notes_rfppRevid && !/^\d+$/.test(input.protectReason_notes_rfppRevid)` or whatever and `alert`s and `return`s if not.  Couldn't hurt?

siddharthvp commented:
> Curious what you think about adding a validation that the revid at least looks right? That is, something in Twinkle.protect.callback.evaluate that does input.protectReason_notes_rfppRevid && !/^\d+$/.test(input.protectReason_notes_rfppRevid) or whatever and alerts and returns if not. Couldn't hurt?

Have you folks looked into [input type=number](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number)? If it triggers browser validations, it should be ideal.

Amorymeltzer commented:
> > Curious what you think about adding a validation that the revid at least looks right? That is, something in Twinkle.protect.callback.evaluate that does input.protectReason_notes_rfppRevid && !/^\d+$/.test(input.protectReason_notes_rfppRevid) or whatever and alerts and returns if not. Couldn't hurt?
> 
> Have you folks looked into [input type=number](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number)? If it triggers browser validations, it should be ideal.

Agreed that'd be nice, although we don't currently support it in Morebits.  I can add that trivially, but as noted, still need/should do on our end anyway.  If memory serves me, IE11 doesn't handle these well.

Amorymeltzer commented:
Opened #1304 to do that @siddharthvp

molly commented:
> Nice, LGTM @molly! Hope I didn't seem like I was insisting on the language, just expressing a preference. You can squash if you like, or I can take care of it, nbd.

Not at all, they were improvements!

> Curious what you think about adding a validation that the revid at least looks right? That is, something in `Twinkle.protect.callback.evaluate` that does `input.protectReason_notes_rfppRevid && !/^\d+$/.test(input.protectReason_notes_rfppRevid)` or whatever and `alert`s and `return`s if not. Couldn't hurt?

Good call, I'll do that.

molly commented:
Added the revision ID validation, and realized it might be a good opportunity to explain how to find the revision ID like we were discussing earlier. I included a link in the alert to https://en.wikipedia.org/wiki/Help:Permanent_link.

Amorymeltzer reviewed (approved):
👍

molly commented:
This is squashed and all ready to go, I think. I can't merge it myself.

Amorymeltzer commented:
Oh yeah, sorry @molly!  I was planning on pushing this up today, actually, as I have a little time I can be available.  Thanks again for the nice work on this!
